Add documentation on testing Chrome Extensions (#7479)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/mv3/automated-testing/index.md

---
layout: "layouts/doc-post.njk"
title: "Automated testing for Chrome Extensions"
seoTitle: "Automated testing for Chrome Extensions"
date: 2023-10-09
description: How to write automated tests for extensions.
---

Automated testing tools provide a great way to regularly test a large amount of your extension's
functionality. This can be useful to ensure key functionality is working and as a way to avoid
regressions between releases.

> ## Unit testing

Author: I like thinking about SEO but I feel like this makes the table of contents hard to parse - thoughts?

<img width="330" alt="Screenshot 2023-10-11 at 10 54 23" src="https://github.com/GoogleChrome/developer.chrome.com/assets/6097064/3895c307-de7f-46ff-8299-ee1e60bf26b2">

> ## Unit testing

Reviewer: Did you consider splitting this into two guides? But you're right, it gets noise if in the same doc.

> ## Unit testing

Author: That's a good idea actually - I'll do that.

of your extension, and outside of the browser.

Code written without using extension APIs can be tested as normal, using a framework such as
[Jest][jest]. To make code easier to test this way, consider using techniques such as
[dependency injection][dependency-injection] which can help to remove dependencies on the chrome
namespace in your lower level implementation.

If you need to test code which includes extension APIs, consider using mocks.

### Example: Using mocks with Jest

Create a `jest.config.js` file, which declares a setup file that will run before all tests:

```js
module.exports = {
  setupFiles: ['<rootDir>/mock-extension-apis.js']
};
```

In `mock-extension-apis.js`, add implementations for the specific functions you expect to call:

```js
global.chrome = {
  tabs: {
    query: async () => { throw new Error("Unimplemented.") };
  }
};
```

Then, use jest.spy to mock a return value in a test:

```js
test("getActiveTabId returns active tab ID", async () => {
  jest.spyOn(chrome.tabs, "query").mockResolvedValue([{
    id: 3,
    active: true,
    currentWindow: true
  }]);
  expect(await getActiveTabId()).toBe(3);
});
```

> ## Integration testing

Reviewer: Also, nit picking. What you're describing is end-to-end testing. Integration testing is when you test the interaction between different components of your system. This does not necessarily include building an extension package. You can (and should) run most of your integration tests outside the browser using mocks etc.

> ## Integration testing

Author: Definitions are hard 😅. I looked around to see what people were using and it seems like the test frameworks that do pick a term use "end-to-end testing" though, so I'll change that up. Thanks for the poke 😁

[Integration testing][integration-testing] involves an extension package being built and loaded into
a browser. A testing tool communicates with the browser to automate interactions and test the same
flows that a user would go through.

See [Testing Chrome Extensions with Puppeteer][tutorial] for a tutorial.

> ### Libraries

Reviewer: ```suggestion
### Using browser testing libraries
```

Extensions are supported by a range of integration testing libraries.

| Library                | Guidance                                                                                                                              |
|------------------------|:-------------------------------------------------------------------------------------------------------------------------------------:|

> | Puppeteer / Playwright | See Chrome Extensions ([Puppeteer][puppeteer-testing] / [Playwright][playwright-testing]).                                            |

Reviewer: Why are both in the same column?

Author: Playwright is a fork of Puppeteer and they generally work very similar (for example, if you compare https://pptr.dev/guides/chrome-extensions and https://playwright.dev/docs/chrome-extensions it is basically the same doc with the header changed).

| Selenium               | Use the [ChromeOptions][selenium-chromeoptions] object to load extensions. More information is available [here][selenium-extensions]. |
| WebDriverIO            | See [Web Extension Testing][webdriverio-testing].                                                                                     |

> ### Headless Chrome

Reviewer: ```suggestion
### Running tests in headless Chrome
```

To be consistent with the other headings

### Testing extension pages

Extension pages can be accessed using their corresponding URL, e.g `chrome-extension://<id>/index.html`. Use the normal navigation methods available in your automation tool of choice to navigate to these URLs.

### Testing an extension popup

Opening an extension popup in the context of another page is not currently possible. Instead, open the popup's URL in a new tab. If your popup uses the active tab, consider implementing an override where a tab ID can be passed explicitly to your popup. For example:

```js
const URL_PARAMS  = new URLSearchParams(window.location.search);

async function getActiveTab() {
  // Open popup.html?tab=5 to use tab ID 5, etc.
  if (URL_PARAMS.has("tab")) {
    return parseInt(URL_PARAMS.get("tab"));
  }

  const tabs = await chrome.tabs.query({
    active: true,
    currentWindow: true
  });

  return tabs[0];
}
```

### Inspecting extension state

> It is generally best practice to avoid accessing internal state in an integration test, and instead you should base your tests on what is visible to the user. However, it can sometimes be desirable to directly access data from the extension. In these cases, consider executing code in the context of an extension page.

Reviewer: ```suggestion
To avoid that test fail, when you change the internal behavior of your extensions, it is generally best practice to avoid accessing internal state in an integration test. Instead you should base your tests on what is visible to the user. However, it can sometimes be desirable to directly access data from the extension. In these cases, consider executing code in the context of an extension page.
```
